clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java

```diff
@@ -204,6 +205,27 @@ private Node getControllerNode(int controllerId, Collection<Node> brokers) {
     }
 
     /**
+     * Returns the set of topics with an error indicating invalid metadata
+     * and topics with any partition whose error indicates invalid metadata.
+     * This includes all non-existent topics specified in the metadata request
+     * and any topic returned with one or more partitions whose leader is not known.
+     */
+    public Collection<String> unavailableTopics() {
+        Collection<String> invalidMetadataTopics = new HashSet<>();
+        for (TopicMetadata topicMetadata : this.topicMetadata) {
+            if (topicMetadata.error.exception() instanceof InvalidMetadataException)
+                invalidMetadataTopics.add(topicMetadata.topic);
+            for (PartitionMetadata partitionMetadata : topicMetadata.partitionMetadata) {
```

Comment:
Does it make sense to put this `for` loop in the `else` block of the `if` above? So if the topic metadata has the exception (and already added to `invalidMetadataTopics`), we no longer look at its partitions?